mlos_bench/mlos_bench/tests/optimizers/grid_search_optimizer_test.py

```diff
@@ -0,0 +1,246 @@
+
+    grid_search_tunables_grid.remove(default_config)
+    assert default_config not in grid_search_opt.pending_configs
+    assert all(config in grid_search_tunables_grid for config in grid_search_opt.pending_configs)
+    assert all(config in list(grid_search_opt.pending_configs) for config in grid_search_tunables_grid)
+
+    # The next suggestion should be a different element in the grid search.
+    suggestion = grid_search_opt.suggest()
+    suggestion_dict = suggestion.get_param_values()
+    assert suggestion_dict != default_config
+    assert suggestion_dict not in grid_search_opt.pending_configs
+    assert suggestion_dict in grid_search_opt.suggested_configs
+    grid_search_opt.register(suggestion, status, score)
+    assert suggestion_dict not in grid_search_opt.pending_configs
+    assert suggestion_dict not in grid_search_opt.suggested_configs
+
+    grid_search_tunables_grid.remove(suggestion.get_param_values())
+    assert all(config in grid_search_tunables_grid for config in grid_search_opt.pending_configs)
+    assert all(config in list(grid_search_opt.pending_configs) for config in grid_search_tunables_grid)
+
+    # FIXME: Should we consider not_converged as the "max_iterations", an empty grid, or both?
+
+    # Try to empty the rest of the grid.
+    while grid_search_opt.not_converged():
+        suggestion = grid_search_opt.suggest()
+        grid_search_opt.register(suggestion, status, score)
+
+    # The grid search should be empty now.
+    assert not list(grid_search_opt.pending_configs)
+    assert not list(grid_search_opt.suggested_configs)
+    assert not grid_search_opt.not_converged()
+
+    # But if we still have iterations left, we should be able to suggest again by refilling the grid.
+    assert grid_search_opt.current_iteration < grid_search_opt.max_iterations
+    assert grid_search_opt.suggest()
+    assert list(grid_search_opt.pending_configs)
+    assert list(grid_search_opt.suggested_configs)
+    assert grid_search_opt.not_converged()
+
+    # Try to finish the rest of our iterations by repeating the grid.
+    while grid_search_opt.not_converged():
+        suggestion = grid_search_opt.suggest()
+        grid_search_opt.register(suggestion, status, score)
+    assert not grid_search_opt.not_converged()
+    assert grid_search_opt.current_iteration >= grid_search_opt.max_iterations
+    assert list(grid_search_opt.pending_configs)
+    assert list(grid_search_opt.suggested_configs)
+
+
+def test_grid_search_async_order(grid_search_opt: GridSearchOptimizer) -> None:
+    """
+    Make sure that grid search optimizer works correctly when suggest and register
+    are called out of order.
+    """
+    score = 1.0
+    status = Status.SUCCEEDED
+    suggest_count = 10
+    suggested = [grid_search_opt.suggest() for _ in range(suggest_count)]
+    suggested_shuffled = suggested.copy()
+    random.shuffle(suggested_shuffled)
+    assert suggested != suggested_shuffled
```

Comment:
```suggestion
```
there's > 0 probability this condition can be false 😃

Reply:
I want it to be true though.  I added a retry loop for now.